torch_xla/csrc/aten_cpu_fallback.cpp

```diff
@@ -1,34 +1,355 @@
+  std::vector<at::Tensor> cuda_tensors(tensors.size());
+  std::transform(tensors.begin(), tensors.end(), cuda_tensors.begin(),
+                 [=](const at::Tensor& tensor) {
+                   // Skip undefined or non-XLA tensors.
+                   if (!is_valid_xla_tensor(tensor)) {
+                     return tensor;
+                   }
+
+                   // Grab the DLManagedTensor.
+                   DLManagedTensor* managed = torch_xla::toDLPack(tensor);
+                   // Remember the device index, so that we can synchronize it
+                   // later. Also ensure that there's only one.
+                   c10::DeviceIndex device =
+                       managed->dl_tensor.device.device_id;
+                   if (*common_device == -1) {
+                     *common_device = device;
+                   } else {
+                     TORCH_CHECK(*common_device == device,
+                                 "fallback using Multi-GPUs not supported");
+                   }
+                   // Create the CUDA tensor.
+                   return at::fromDLPack(
+                       managed, [=](void*) { managed->deleter(managed); });
+                 });
+  return cuda_tensors;
+}
+
+// Copy back the results from CUDA to XLA.
+// Assumes that we have already synchronized CUDA.
+static at::Tensor to_xla(const at::Tensor& tensor) {
+  return torch_xla::fromDLPack(at::toDLPack(tensor));
+}
+
+// Synchronizes the CUDA device being used by PyTorch.
+static void torch_cuda_synchronize(c10::DeviceIndex common_device) {
+  // Save the current PyTorch device, in case it's not the same as the
+  // recorded tensor device.
+  c10::DeviceIndex current = c10::cuda::current_device();
+  c10::cuda::set_device(common_device);
+  c10::cuda::device_synchronize();
+  c10::cuda::set_device(current);
+}
+
+// Former 'cpu_fallback'.
+// Changes:
+//
+//   1. Track the device index being used. Rationale: we synchronize the device
+//      before crossing device borders for correctness.
+//
+void cuda_fallback(const c10::OperatorHandle& op, torch::jit::Stack* stack,
+                   bool error_on_views) {
+  auto& schema_args = op.schema().arguments();
+  const auto num_arguments = schema_args.size();
+  auto arguments = torch::jit::last(stack, num_arguments);
+  const auto arguments_begin = stack->size() - num_arguments;
+
+  std::vector<at::Tensor> tensor_args;
+  std::vector<int> tensor_args_indices;
+
+  std::vector<c10::List<at::Tensor>> tensorlist_args;
+  std::vector<int> tensorlist_args_indices;
+
+  std::vector<c10::IValue> tensorlist_cuda_args;
+
+  // This fallback only works if all XLA:CUDA tensor arguments are
+  // on the same CUDA device.
+  //
+  // We keep track of said device, so that after actually running
+  // the operation on PyTorch CUDA eager-mode, we synchronize the
+  // device.
+  //
+  // This variable is updated over the course of 'to_cuda' calls.
+  c10::DeviceIndex common_device = -1;
```

Comment:
I'm confused about the `common_device`. Do we ever change the value after this line?

Reply:
Would the original `tgt_device` be easier to understand?

Reply:
We do change inside `to_cuda` function. There, we check the device of every XLA tensor, so that we are able to synchronize the computation later.

> Would the original `tgt_device` be easier to understand?

Since they have different types, and are used in different ways, I thought that it would be a bit confusing to name it `tgt_device`.

Reply:
I can see your point. Maybe add some comment such as: common_device refers to the device that all tensors should be on; Ideally, all the tensors should be on the same device

Reply:
Just did that.

```c++
  // Common device for all XLA tensors.
  //
  // CUDA OpenXLA fallback is supported only when all XLA tensors live in
  // the same XLA device. This field should be updated and checked every
  // time we convert an XLA tensor argument into a CUDA tensor.
  c10::Device common_device;
```